src/connector/src/sink/mongodb.rs

```diff
@@ -0,0 +1,750 @@
+        if self.pk_indices.len() > 1
+            || self.schema.fields[self.pk_indices[0]].name != MONGODB_PK_NAME
+        {
+            // compound pk should not have a field named `_id`
+            document.insert(MONGODB_PK_NAME, pk.clone());
+        }
+
+        let pk = doc! {MONGODB_PK_NAME: pk};
+        match op {
+            Op::Insert | Op::UpdateInsert => match upsert_builder.entry(ns) {
+                Entry::Occupied(mut entry) => entry.get_mut().add_upsert(pk, document),
+                Entry::Vacant(entry) => {
+                    let mut builder = UpsertCommandBuilder::new(coll, self.max_entries);
+                    builder.add_upsert(pk, document)?;
+                    entry.insert(builder);
+                    Ok(())
+                }
+            },
+            Op::UpdateDelete => Ok(()),
+            Op::Delete => match upsert_builder.entry(ns) {
+                Entry::Occupied(mut entry) => entry.get_mut().add_delete(pk),
+                Entry::Vacant(entry) => {
+                    let mut builder = UpsertCommandBuilder::new(coll, self.max_entries);
+                    builder.add_delete(pk)?;
+                    entry.insert(builder);
+                    Ok(())
+                }
+            },
+        }
+    }
+
+    async fn write(&mut self, op: Op, row: RowRef<'_>) -> Result<()> {
+        let mut command_builder = self.command_builder.take();
+        match command_builder {
+            Some(CommandBuilder::AppendOnly(ref mut insert_builder)) => {
+                if op != Op::Insert {
+                    if let Ok(suppressed_count) = LOG_SUPPERSSER.check() {
+                        tracing::warn!(
+                            suppressed_count,
+                            ?op,
+                            ?row,
+                            "non-insert op received in append-only mode"
+                        );
+                    }
+                    return Ok(());
+                }
+                self.append(insert_builder, row)?;
+            }
+            Some(CommandBuilder::Upsert(ref mut upsert_builder)) => {
+                if op == Op::UpdateDelete {
+                    // we should ignore the `UpdateDelete` in upsert mode
+                    return Ok(());
+                }
+                self.upsert(upsert_builder, op, row)?;
+            }
+            _ => unreachable!(),
+        }
+        self.command_builder = command_builder;
+
+        self.buffered_entries += 1;
+        if self.buffered_entries >= self.max_entries {
+            self.flush().await?;
+        }
+
+        Ok(())
+    }
+
+    async fn flush(&mut self) -> Result<()> {
+        let mut command_builder = self.command_builder.take();
+        match command_builder {
+            Some(CommandBuilder::AppendOnly(ref mut insert_builder)) => {
+                for (ns, builder) in insert_builder.drain() {
```

Comment:
May add a `TODO` here for the previously discussed comment about concurrent send_bulk_write_command.